sql/parser/builtins.go

```diff
@@ -943,6 +961,108 @@ func datumToString(datum Datum) (string, error) {
 	return "", fmt.Errorf("argument type unsupported: %s", datum.Type())
 }
 
+var replaceSubRe = regexp.MustCompile(`\\[&1-9]`)
+
+func regexpReplace(s, pattern, to, sqlFlags string) (Datum, error) {
+	pattern, global, err := regexpEvalFlags(pattern, sqlFlags)
+	if err != nil {
+		return nil, err
+	}
+
+	patternRe, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+
+	matchCount := 1
+	if global {
+		matchCount = -1
+	}
+
+	replaceIndex := 0
+	var newString bytes.Buffer
+
+	// regexp.ReplaceAllStringFunc cannot be used here because it does not provide
+	// access to regexp submatches for expansion in the replacement string.
+	// regexp.ReplaceAllString cannot be used here because it does not allow
+	// replacement of a specific number of matches, and does not expose the full
+	// match for expansion in the replacement string.
+	for _, matchIndex := range patternRe.FindAllStringSubmatchIndex(s, matchCount) {
+		start := matchIndex[0]
+		end := matchIndex[1]
+
+		preMatch := s[replaceIndex:start]
+		newString.WriteString(preMatch)
+
+		match := s[start:end]
+		matchTo := replaceSubRe.ReplaceAllStringFunc(to, func(repl string) string {
+			subRef := repl[len(repl)-1]
+			if subRef == '&' {
+				return match
+			}
+
+			sub, _ := strconv.Atoi(string(subRef))
+			if 2*sub >= len(matchIndex) {
+				return ""
+			}
+
+			subStart := matchIndex[2*sub]
+			subEnd := matchIndex[2*sub+1]
+			return s[subStart:subEnd]
+		})
+		newString.WriteString(matchTo)
+
+		replaceIndex = end
+	}
+	newString.WriteString(s[replaceIndex:])
+
+	return DString(newString.String()), nil
+}
+
+// regexpEvalFlags evaluates the provided Postgres regexp flags in
+// accordance with their definitions provided at
+// http://www.postgresql.org/docs/9.0/static/functions-matching.html#POSIX-EMBEDDED-OPTIONS-TABLE.
+// It then returns an adjusted regexp pattern, and a flag specifying if more
+// than one match should be found or not.
+func regexpEvalFlags(pattern, sqlFlags string) (string, bool, error) {
+	global := false
+	goReFlags := map[rune]bool{'s': true, 'm': true}
```

Comment:
Interestingly code reviewers seemed to push for the `map[T]bool` method of implementing a _set_ inside Google, stressing ease of use (`if map[var] { ... }`) over efficiency. I personally like the empty struct way better because there are only two states available (in map & out of map) instead of (in map true, in map false, & out of map), so if that's how we do it here, I'll happily make the change.